src/core/pointcloud/expression/qgspointcloudexpression.h

```diff
@@ -0,0 +1,263 @@
+     * Create a copy of this expression. This is preferred
+     * over recreating an expression from a string since
+     * it does not need to be re-parsed.
+     */
+    QgsPointCloudExpression( const QgsPointCloudExpression &other );
+
+    /**
+     * Create a copy of this expression. This is preferred
+     * over recreating an expression from a string since
+     * it does not need to be re-parsed.
+     */
+    QgsPointCloudExpression &operator=( const QgsPointCloudExpression &other );
+
+    /**
+     * Automatically convert this expression to a string where requested.
+     */
+    operator QString() const SIP_SKIP;
+
+    /**
+     * Create an empty expression.
+     */
+    QgsPointCloudExpression();
+
+    ~QgsPointCloudExpression();
+
+    /**
+     * Compares two expressions. The operator returns TRUE
+     * if the expression string is equal.
+     */
+    bool operator==( const QgsPointCloudExpression &other ) const;
+
+    /**
+     * Checks if this expression is valid.
+     * A valid expression could be parsed but does not necessarily evaluate properly.
+     */
+    bool isValid() const;
+
+    /**
+     * Returns TRUE if an error occurred when parsing the input expression
+     */
+    bool hasParserError() const;
+
+    /**
+     * Returns parser error
+     */
+    QString parserErrorString() const;
+
+    /**
+     * Returns parser error details including location of error.
+     */
+    QList<QgsExpression::ParserError> parserErrors() const;
+
+    /**
+     * Returns the root node of the expression.
+     *
+     * The root node is NULLPTR if parsing has failed.
+     */
+    const QgsPointCloudExpressionNode *rootNode() const;
+
+    /**
+     * Gets the expression ready for evaluation.
+     * \param block pointer to the QgsPointCloudBlock that will be filtered
+     */
+    bool prepare( const QgsPointCloudBlock *block );
+
+    /**
+     * Gets list of attributes referenced by the expression.
+     */
+    QSet<QString> referencedAttributes() const;
+
+#ifndef SIP_RUN
+
+    /**
+     * Returns a list of all nodes which are used in this expression
+     *
+     * \note not available in Python bindings
+     */
+    QList<const QgsPointCloudExpressionNode *> nodes( ) const;
+
+    /**
+     * Returns a list of all nodes of the given class which are used in this expression
+     *
+     * \note not available in Python bindings
+     */
+    template <class T>
+    QList<const T *> findNodes( ) const
+    {
+      QList<const T *> lst;
+      const QList<const QgsPointCloudExpressionNode *> allNodes( nodes() );
+      for ( const auto &node : allNodes )
+      {
+        const T *n = dynamic_cast<const T *>( node );
+        if ( n )
+          lst << n;
+      }
+      return lst;
+    }
+#endif
+
+    // evaluation
+
+    /**
+     * Evaluate the expression for one point.
+     * \returns 0.0 for false or 1.0 for true.
+     * \param p point number within the block to evaluate
+     */
+    double evaluate( int p );
```

Comment:
Let's rename `p` here as well.

It would be good to add to the docstring that expression has to be already prepared

Reply:
done